Windows fixes & installation helpers

rok-cesnovar opened the pull request:
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

- fixes #20 
- fixes #259 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
Rok Češnovar, Uni. of Ljubljana


By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

jgabry commented on R/install.R:
```diff
@@ -396,3 +396,142 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+    # On Unix systems we check for make and a suitable compiler
+    make_path <- dirname(Sys.which("make"))
+    if (!nzchar(make_path)) {
+      stop("The 'make' tool was not found. Please install 'make' and restart the R session.", call. = FALSE)
```
I thought that `xcode-select --install` was enough but I think at one point @mitzimorris found that she had to install all of xcode. Mitzi am I remembering that correctly? So to be safe we can recommend installing all of xcode I guess. 

To detect if a user is on mac I think this should work:

```r
Sys.info()[["sysname"]] == "Darwin"
```

mitzimorris replied:
regarding xcode - yes, I saw this during an upgrade to Catalina - under some circumstances this might be necessary - wish I understood which circumstance that is.   I tested a few different things - upgrade, upgrade + migration - on Bob's old 2011 MacBookPro - too many variables there.

jgabry commented on R/install.R:
```diff
@@ -396,3 +396,142 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+    gpp_path <- dirname(Sys.which("g++"))
+    clang_path <- dirname(Sys.which("clang++"))
+    if (!nzchar(gpp_path) && !!nzchar(clang_path)) {
+      stop("A C++ compiler was not found. Please install the 'clang++' of 'g++' compiler and restart the R session.", call. = FALSE)
```
I think it's probably OK to just error on Mac. On Linux there's probably more of a chance of someone trying a different compiler so maybe a warning would be preferable. That said, I'd also be ok with an error and requiring the officially supported compiler.

jgabry commented on R/install.R:
```diff
@@ -396,3 +396,142 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+  invisible(NULL)
+}
+
+install_cmdstan_toolchain <- function(check_only = FALSE) {
```
Yeah running it with `check_only=TRUE` seems like a good idea.

rok-cesnovar commented:
This now has all the functionalities it will have in the end. Its only missing tests for which I haven't completely thought out how to do. 

Current functionalities:
- `check_cmdstan_toolchain(fix = FALSE)` checks for make and g++/clang++
- `check_cmdstan_toolchain(fix = TRUE)` will fix what it can fix on Windows (install mingw32-make and fix PATH issues)
- `install_cmdstan()` calls `check_cmdstan_toolchain(fix = FALSE)`, but you can turn it off with check_toolchain = FALSE

The checks are: 

**Windows R4.0:** 
- is Rtools 4.0 installed?
- is RTools 4.0 installed in a valid path?
- are mingw32-make and g++ available?
- are mingw32-make and g++ that are picked up the ones in RTools 4.0?

**Windows R3.x:** 
- are mingw32-make and g++ available?
- if above is FALSE check a bunch of typical RTools 3.5 paths

**Unix:**
- check if make and g++ are available

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+          if (!fix) {
+            error_message <- c(
+              "",
+              "A RTools installation was found, but the PATH is not properly set.",
```
For consistency with the same message above this should be 
```suggestion
              "RTools installation found but PATH was not properly set.",
```

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+            error_message <- c(
+              "",
+              "A RTools installation was found, but the PATH is not properly set.",
+              "Run 'cmdstanr::check_cmdstan_toolchain(check_only = FALSE)' to fix the issue."
```
```suggestion
              "Run check_cmdstan_toolchain(fix = TRUE) to fix the issue."
```

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+            )
+            stop(paste0(error_message, collapse = "\n"), call. = FALSE)
+          }
+          if (!quiet) message("Writing RTools path to ~/.Renviron ...")
```
Nitpicking for style consistency: 

```suggestion
          if (!quiet) {
            message("Writing RTools path to ~/.Renviron ...")
          }
```

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+            write(paste0('RTOOLS35_HOME=', rtools_path), file = "~/.Renviron", append = TRUE)
+          }
+          write('PATH="${RTOOLS35_HOME}\\bin;${RTOOLS35_HOME}\\mingw_64\\bin;${PATH}"', file = "~/.Renviron", append = TRUE)
+          stop("Restart R and run check_cmdstan_toolchain(check_only = TRUE) to confirm the installation was successful.", call. = FALSE)
```
```suggestion
          stop("Please restart R and run check_cmdstan_toolchain() to confirm the installation was successful.", call. = FALSE)
```

But since this same message is used many times maybe it's worth just making a helper function that returns this error message?

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+          stop("Restart R and run check_cmdstan_toolchain(check_only = TRUE) to confirm the installation was successful.", call. = FALSE)
+        } else {
+          error_message <- c(
+            "",
+            "A toolchain was not found. Please install RTools 3.5 and run",
+            "",
+            "write(\'RTOOLS35_HOME=rtools35/install/path/\', file = \"~/.Renviron\", append = TRUE)",
+            "replacing 'rtools35/install/path/' with the actual install path of RTools 3.5.",
+            "",
+            "Then restart R and run 'cmdstanr::check_cmdstan_toolchain(check_only = FALSE)'."
+          )
+          stop(paste0(error_message, collapse = "\n"), call. = FALSE)
```
Oh, now I think I see why you're using the blank `""` in most of these error messages. If you prefer this way that's fine, but you can get the same output with fewer lines of code like this: 

```
stop(
  "A toolchain was not found. Please install RTools 3.5 and run",
  "\n\nwrite(\'RTOOLS35_HOME=rtools35/install/path/\', file = \"~/.Renviron\", append = TRUE)",
  "\nreplacing 'rtools35/install/path/' with the actual install path of RTools 3.5.",
  "\n\nThen restart R and run 'cmdstanr::check_cmdstan_toolchain(check_only = FALSE)'.", 
  call. = FALSE
)
```

In either case, `'cmdstanr::check_cmdstan_toolchain(check_only = FALSE)'` can be replaced with 
`check_cmdstan_toolchain(fix = TRUE)`

rok-cesnovar replied:
Thanks, replaced with your suggestion.

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+    make_path <- dirname(Sys.which("make"))
+    if (!nzchar(make_path)) {
+      if (os_is_macos()) {
+        stop("The 'make' tool was not found. Please install 'xcode' toolchains with 'xcode-select --install' or the app store and restart the R session.", call. = FALSE)
```
```suggestion
        stop("The 'make' tool was not found. Please install the command line tools for Mac with 'xcode-select --install' or install Xcode from the app store. Then restart R and run check_cmdstan_toolchain().", call. = FALSE)
```

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+      if (os_is_macos()) {
+        stop("The 'make' tool was not found. Please install 'xcode' toolchains with 'xcode-select --install' or the app store and restart the R session.", call. = FALSE)
+      } else {
+        stop("The 'make' tool was not found. Please install 'make' and restart the R session.", call. = FALSE)
```
```suggestion
        stop("The 'make' tool was not found. Please install 'make', restart R, and then run check_cmdstan_toolchain().", call. = FALSE)
```

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+      }
+    }
+  }
+  if (!quiet) message("The CmdStan toolchain is setup properly!")
```
Nitpicking for code style consistency: 

```suggestion
  if (!quiet) {
    message("The CmdStan toolchain is setup properly!")
  }
```

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+    clang_path <- dirname(Sys.which("clang++"))
+    if (!nzchar(gpp_path) && !nzchar(clang_path)) {
+      if (os_is_macos()) {
+        stop("A suitable C++ compiler was not found. Please install 'xcode' toolchains with 'xcode-select --install' or the app store and restart the R session.", call. = FALSE)
```
```suggestion
        stop("A suitable C++ compiler was not found. Please install the command line tools for Mac with 'xcode-select --install' or install Xcode from the app store. Then restart R and run check_cmdstan_toolchain().", call. = FALSE)
```

jgabry commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+      if (os_is_macos()) {
+        stop("A suitable C++ compiler was not found. Please install 'xcode' toolchains with 'xcode-select --install' or the app store and restart the R session.", call. = FALSE)
+      } else {
+        stop("A C++ compiler was not found. Please install the 'clang++' or 'g++' compiler and restart the R session.", call. = FALSE)
```
```suggestion
        stop("A C++ compiler was not found. Please install the 'clang++' or 'g++' compiler, restart R, and run check_cmdstan_toolchain().", call. = FALSE)
```

rok-cesnovar commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+      # we assume that RTools 40 is not installed.
+      if (!nzchar(rtools_path)) {
+        error_message <- c(
+          "",
```
Yeah, to make newlines. Fixed by using "\n" directly.

rok-cesnovar commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+        stop(paste0(error_message, collapse = "\n"), call. = FALSE)
+      }
+      if (fix) {
+        install_mingw32_make()
```
It wont reinstall mingw32-make if its already installed and no updated versions are available, which I think is fine. Its also quick in that case.

Moved it now to only run if mingw32-make/g++ are not found or the wrong mingw32-make/g++ are picked up.

rok-cesnovar commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+            file.path("c:/", "rtools"),
+            file.path("c:/", "rtools35")
+          )
+          for (p in typical_paths) {
+            if (dir.exists(p)) {
+              rtools_path <- p
```
Good question. Moved to erroring if multiple found.

rok-cesnovar commented on R/install.R:
```diff
@@ -396,3 +398,172 @@ build_status_ok <- function(process_log, quiet = FALSE) {
+  invisible(NULL)
+}
+
+check_compiler_version <- function(compiler = c("g++", "clang++")) {
+
+}
```
At first I meant to define it but I later figured it will be challenging, because we cant really know whether the user uses g++ or clang before building cmdstan. We should add a compiler check command to cmdstan makefiles. 
We can use that then. 

Too old compiler is generally not a huge support problem. Nowadays its mostly seen when users move to a cluster, where some popular distributions still come with g++ 4.8.x. But users mostly run cmdstan directly on a cluster.

rok-cesnovar commented:
> Then `check_cmdstan_toolchain()` can call those functions and it would be much easier to tell what `check_cmdstan_toolchain()` is doing. Does that make sense? If so, what do you think?

Completely agree. Split the function now.

Also addressed all your other comments and applied the suggested fixes. Thanks for those! They were really helpful.

This is now ready in terms of functionalities, though I wont to go over it again (I have AWS instances with various setups to test this). I would also like to add a few unit tests where it's possible. It's not possible for all cases. Will add those Monday/Tuesday and will then tag you for another review.

jgabry commented:
Ok great, sounds good. Thanks!

rok-cesnovar commented:
Note to self: check that we have write access to the Rtools folder before trying to install mingw32-make.

jgabry commented:
@rok-cesnovar I see this is passing all the checks now. Is it ready for another review or are you still working on it? No rush, just checking.

rok-cesnovar commented:
Still missing tests. Sorry, still haven't gotten back to this.

jgabry commented:
> Sorry, still haven't gotten back to this.

No worries!

jgabry commented:
And definitely no rush if you’ve got a lot going on. This can wait!

rok-cesnovar commented:
This is now ready for merging. I added some tests.

jgabry commented on R/install.R:
```diff
@@ -48,6 +52,9 @@
 #' @param cpp_options A list specifying any makefile flags/variables to be
 #'   written to the `make/local` file. For example, `list("CXX" = "clang++")`
 #'   will force the use of clang for compilation.
+#' @param check_toolchain A list specifying any makefile flags/variables to be
```
Just noticed that the doc is wrong here. I think it's just a copy/paste error. check_toolchain should be T/F right?

rok-cesnovar replied:
Indeed. A copy paste error. Fixed now. Thanks!

jgabry commented on R/install.R:
```diff
@@ -30,8 +30,12 @@
+#'   The default is `FALSE`.
+#'   For `check_cmdstan_toolchain()`, should the function supress 
+#'   printing informational messages? The default is `FALSE`.
+#'   If `TRUE` `check_cmdstan_toolchain()` only output errors.
```
```suggestion
#'   If `TRUE` `check_cmdstan_toolchain()` only outputs errors.
```

jgabry reviewed (commented):
Awesome, thanks for adding the tests.  Found a few tiny doc things, but then merging is fine by me. This will be great to have.

rok-cesnovar commented:
Thanks! Will merge once tests pass.
